test/hooks/useSelector.spec.tsx

```diff
@@ -660,6 +640,69 @@ describe('React', () => {
           expect(renderedItems.length).toBe(2)
           expect(renderedItems[0]).toBe(renderedItems[1])
         })
+
+        it('should have linear or better unsubscribe time, not quadratic', () => {
+          const reducer = (state: number = 0, action: any) =>
+            action.type === 'INC' ? state + 1 : state
+          const store = createStore(reducer)
+          const increment = () => ({ type: 'INC' })
+
+          const numChildren = 100000
+
+          function App() {
+            useSelector((s: number) => s)
+            const dispatch = useDispatch()
+
+            const [children, setChildren] = useState(numChildren)
+
+            const toggleChildren = () =>
+              setChildren((c) => (c ? 0 : numChildren))
+
+            return (
+              <div>
+                <button onClick={toggleChildren}>Toggle Children</button>
+                <button onClick={() => dispatch(increment())}>Increment</button>
+                {[...Array(children).keys()].map((i) => (
+                  <Child key={i} />
+                ))}
+              </div>
+            )
+          }
+
+          function Child() {
+            useSelector((s: number) => s)
+            // Deliberately do not return any DOM here - we want to isolate the cost of
+            // unsubscribing, and tearing down thousands of JSDOM nodes is expensive and irrelevant
+            return null
+          }
+
+          const { getByText } = rtl.render(
+            <ProviderMock store={store}>
+              <App />
+            </ProviderMock>
+          )
+
+          const timeBefore = Date.now()
+
+          const button = getByText('Toggle Children')
+          rtl.act(() => {
+            rtl.fireEvent.click(button)
+          })
+
+          const timeAfter = Date.now()
+          const elapsedTime = timeAfter - timeBefore
+
+          // Seeing an unexpected variation in elapsed time between React 18 and React 17 + the compat entry point.
+          // With 18, I see around 75ms with correct implementation on my machine, with 100K items.
+          // With 17 + compat, the same correct impl takes about 4200-5000ms.
+          // With the quadratic behavior, this is at least 13000ms (or worse!) under 18, and 22000ms+ with 17.
+          // The 13000ms time for 18 stays the same if I use the shim, so it must be a 17 vs 18 difference somehow,
+          // although I can't imagine why, and if I remove the `useSelector` calls both tests drop to ~50ms.
+          // So, we'll modify our expectations here depending on whether this is an 18 or 17 compat test,
+          // and give some buffer time to allow for variations in test machines.
+          const expectedMaxUnmountTime = IS_REACT_18 ? 500 : 7000
+          expect(elapsedTime).toBeLessThan(expectedMaxUnmountTime)
```

Comment:
I do worry about the stability of this. This seems like a future case of "works on my machine!" being a false-positive result. 

I don't think so based on how things are set up, but would it be possible to compare the React 17 vs 18 times to have an acceptable ratio instead? That would maintain stability between different CPU speeds.

Reply:
I don't particularly like it either :)  But the "incorrect" times I was seeing were 13.5s for 18 and 23s for 17, and given the number of children being unmounted I see that as a big enough gap that I don't think we're likely to have a false positive due to machine/environment variance.

I do kinda want to investigate _why_ it's slower with React 17 a bit further. But the immediate goal for this test is just to specifically document that "we expect unsubscribing lots of components to not be quadratic" so that we don't forget in the future, and catch that if we were to ever somehow lose that behavior.  I think the current test at least accomplishes that.

Reply:
Worst case, the test randomly starts failing due to CPU/env stuff, and we change how it works then.